File: server/intelligence/news-extractor-service.ts
/**
 * Python news extraction service wrapper
 * Uses newspaper4k for robust article extraction
 */

import { exec } from "child_process";
import { promisify } from "util";
import path from "path";

const execAsync = promisify(exec);

const PYTHON_SCRIPT = path.join(process.cwd(), "news-extractor-service.py");

export interface ExtractedArticle {
  title: string;
  text: string;
  authors: string[];
  publish_date?: string;
  summary: string;
  keywords: string[];
  url: string;
  publishedAt?: string;
  description?: string;
  source?: string;
}

/**
 * Extract article content using Python newspaper4k service
 */
export async function extractArticleWithNewspaper4k(
  url: string
): Promise<ExtractedArticle | null> {
  try {
    const { stdout, stderr } = await execAsync(
      `python3 ${PYTHON_SCRIPT} --url "${url}"`
    );

    if (stderr && !stderr.includes("Error")) {
      console.warn(`[News Extractor] Warning: ${stderr}`);
    }

    const result = JSON.parse(stdout);
    return result;
  } catch (error: any) {
    console.error(`[News Extractor] Failed to extract ${url}:`, error.message);
    return null;
  }
}

/**
 * Fetch articles from RSS feed using Python service
 */
export async function fetchRSSWithNewspaper4k(
  feedUrl: string,
  daysBack: number = 90
): Promise<ExtractedArticle[]> {
  try {
    const { stdout, stderr } = await execAsync(
      `python3 ${PYTHON_SCRIPT} --rss "${feedUrl}" --days ${daysBack}`
    );

    if (stderr && !stderr.includes("Error")) {
      console.warn(`[News Extractor] Warning: ${stderr}`);
    }

    const articles = JSON.parse(stdout);
    return Array.isArray(articles) ? articles : [];
  } catch (error: any) {
    console.error(
      `[News Extractor] Failed to fetch RSS ${feedUrl}:`,
      error.message
    );
    return [];
  }
}

/**
 * Check if Python service is available
 */
// Cache the Python service check result to avoid repeated checks
let pythonServiceAvailable: boolean | null = null;
let pythonServiceChecked = false;

export async function checkPythonService(): Promise<boolean> {
  // Only check once per process
  if (pythonServiceChecked) {
    return pythonServiceAvailable || false;
  }

  pythonServiceChecked = true;

  try {
    await execAsync(`python3 --version`);
    // Check if newspaper4k is installed (newspaper4k uses 'newspaper' module name)
    try {
      const { stdout, stderr } = await execAsync(
        `python3 -c "from newspaper import Article; print('OK')" 2>&1`
      );
      if (stdout.includes('OK') || stdout.trim() === 'OK') {
        pythonServiceAvailable = true;
        console.log("[News Extractor] Python newspaper4k service is available");
        return true;
      }
      pythonServiceAvailable = false;
      return false;
    } catch (error: any) {
      // Check stderr for import errors
      const errorOutput = error.stderr || error.stdout || error.message || "";
      if (errorOutput.includes("ModuleNotFoundError") || errorOutput.includes("ImportError")) {
        if (!(global as any).__python_service_warned) {
          console.warn(
            "[News Extractor] newspaper4k not installed. Run: pip3 install -r requirements-news.txt"
          );
          (global as any).__python_service_warned = true;
        }
      }
      pythonServiceAvailable = false;
      return false;
    }
  } catch {
    if (!(global as any).__python_service_warned) {
      console.warn("[News Extractor] Python3 not found");
      (global as any).__python_service_warned = true;
    }
    pythonServiceAvailable = false;
    return false;
  }
}

File: news-extractor-service.py
#!/usr/bin/env python3
"""
News extraction microservice using newspaper4k
Provides robust article extraction that's proven and maintained
"""

import sys
import json
import argparse
from newspaper import Article
from datetime import datetime, timedelta
import requests
from typing import List, Dict, Optional

def extract_article(url: str) -> Optional[Dict]:
    """Extract article content using newspaper4k"""
    try:
        article = Article(url)
        article.download()
        article.parse()
        
        return {
            "title": article.title or "",
            "text": article.text or "",
            "authors": article.authors or [],
            "publish_date": article.publish_date.isoformat() if article.publish_date else None,
            "summary": article.summary or "",
            "keywords": article.keywords or [],
            "url": url,
        }
    except Exception as e:
        print(f"Error extracting {url}: {e}", file=sys.stderr)
        return None

def fetch_rss_articles(feed_url: str, days_back: int = 90) -> List[Dict]:
    """Fetch articles from RSS feed and extract full content"""
    try:
        import feedparser
        feed = feedparser.parse(feed_url)
        articles = []
        cutoff_date = datetime.now() - timedelta(days=days_back)
        
        for entry in feed.entries[:50]:  # Limit to 50 most recent
            url = entry.get('link', '')
            if not url:
                continue
            
            # Check date if available
            pub_date = None
            if hasattr(entry, 'published_parsed') and entry.published_parsed:
                pub_date = datetime(*entry.published_parsed[:6])
            elif hasattr(entry, 'updated_parsed') and entry.updated_parsed:
                pub_date = datetime(*entry.updated_parsed[:6])
            
            # Include if date is missing or within window
            if pub_date is None or pub_date >= cutoff_date:
                article_data = extract_article(url)
                if article_data:
                    article_data["publishedAt"] = pub_date.isoformat() if pub_date else datetime.now().isoformat()
                    article_data["description"] = entry.get('summary', article_data.get('summary', ''))
                    article_data["source"] = feed.feed.get('title', '')
                    articles.append(article_data)
        
        return articles
    except Exception as e:
        print(f"Error fetching RSS {feed_url}: {e}", file=sys.stderr)
        return []

def main():
    parser = argparse.ArgumentParser(description='News extraction service')
    parser.add_argument('--url', help='Extract single article URL')
    parser.add_argument('--rss', help='Fetch from RSS feed URL')
    parser.add_argument('--days', type=int, default=90, help='Days back to fetch')
    
    args = parser.parse_args()
    
    if args.url:
        result = extract_article(args.url)
        if result:
            print(json.dumps(result))
    elif args.rss:
        articles = fetch_rss_articles(args.rss, args.days)
        print(json.dumps(articles))
    else:
        print(json.dumps({"error": "Must provide --url or --rss"}), file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

